modules/crypto/build.gradle

```diff
@@ -15,12 +15,17 @@ thirdPartyAudit.enabled = false
 forbiddenApisTest.ignoreFailures = true
 testingConventions.enabled = false
 
+opensearchplugin {
```

Comment:
@vikasvb90 please move it to plugins (or even better, sandbox plugins), we should not install it by default, this is experimental feature

Reply:
Thanks @reta  . Users need to provide `CryptoMetadata` while registering a repository. Hence this feature wouldn't get used by default.  The default installation would only populate the `CryptoHandlerRegistry` with one option of doing enc-dec . If we install it as a module , it makes usage  easier by just supplying/installing the KeyProvider  and not bothering about installing/configuring `CryptoModulePlugin`.

Reply:
Thanks @gbbafna 

>  Hence this feature wouldn't get used by default. 

Modules supposed to have the functionality that is critical for OpenSearch to function, since this feature is optional (from above), the convince to use is not an argument to bundle it in. We could reconsider the decision(s) later on if needed.

Reply:
In the same vein, this is experimental API and feature, why it is not behind the experimental flag?

Reply:
@reta By this line ` since this feature is optional (from above), the convince to use is not an argument to bundle it in` are you saying that if there is a module then it becomes mandatory to be used in OpenSearch and it can't be a user's choice? I was looking at `repository-url` module and not sure if that fits your theory. Correct me pls if I am wrong!

Reply:
>  are you saying that if there is a module then it becomes mandatory to be used in OpenSearch and it can't be a user's choice? I 

@vikasvb90 modules are installed automatically with each OpenSearch distribution, whereas plugins are bundled but not installed.

Reply:
I am referring specifically to your point that a module once added to OpenSearch can't be disable or used optionally by a user. By this logic, `repository-url` should necessarily get used by all users. Right?

Reply:
Intent here was to not treat it as optional or non-optional feature but as two flavors of data e.g. a repository. And that's why I think the decision was taken by @andrross  and @Bukhtawar . I will let others comment on this. But since the decision was already made, I would still prefer to push it now and then refactor later as we are a bit late to make the structural change.

Reply:
@andrross @Bukhtawar I was reading through the discussion https://github.com/opensearch-project/OpenSearch/pull/8465#discussion_r1294940922 and I have not found any strong arguments toward making a module to support the cryptography (the name `crypto` is heavenly overloaded), any objections to have it as plugin since at this moment, it is not required and is purely opt-in

Reply:
Also, we want security to be first class citizen in OpenSearch and even preferred core for our changes . Due to issues in security plugin , we had to move this to module .

Reply:
@reta This isn't implementing the feature itself. The actual feature would have to be implemented by a concrete plugin implementation and hence I don't see much value with this logic to be made experimental as long as you see an upfront risk with this dependency which might cause regression with the existing code base.

Moving this to plugin would require two plugins to be installed which isn't the optimal experience IMO. Trying to understand your concern better

Reply:
Thank @Bukhtawar 

> This isn't implementing the feature itself. The actual feature would have to be implemented by a concrete plugin implementation 

Now I am somewhat confused what is the role of this module? So we do have `CryptoPlugin` which is (to me) the extension point that has to be implemented by concrete plugin. This module is just another CryptoPlugin? Also, my biggest concern is module is apparently not tested (you could easily tell that by checking that there is no security policy and Bouncycastle needs tons of permissions).

Reply:
`my biggest concern is module is apparently not tested` This has been perf tested. No jsm permission is required by AWS encryption SDK for it to be used in encryption/decryption. The blank which is nothing but NoOpCryptoHandler was added later as the review from crypto team is pending over partial encryption changes.

Reply:
> This has been perf tested

@vikasvb90 how this plugin / module have been tested if this is the pull request that introduces it first time?

Reply:
@reta I replied to this `there is no security policy and Bouncycastle needs tons of permissions`. I have perf tested this in lib and it didn't require any BC jsm permission.  I am about to test the restructured changes as well with actual CryptoHandler which does actual encryption/decryption.

Reply:
Just recreated a draft PR again from previous commits - https://github.com/opensearch-project/OpenSearch/pull/9823 for your reference. This has actual encryption/decryption impl.

Reply:
The vision here is this module itself would contain all types of encryption handlers - frame based , block based , CTR  .  Depending upon the settings, a cluster would be able to chose its own method of encryption  .

Reply:
@vikasvb90 thank you, since more details on the matter are revealed, I am more on the side to not even have changes delivered in 2.10 but move it off to 2.11 when with actual CryptoHandler which does actual encryption/decryption. If we want to show case uses this new feature, it is valid point - it should go to `plugins/example-plugins` instead, definitely (to me) not in a state to be promoted to a core  module.

Reply:
@gbbafna thank you

> The vision here is this module itself would contain all types of encryption handlers - frame based , block based , CTR . Depending upon the settings, a cluster would be able to chose its own method of encryption .

Could we introduce this module / plugin / lib when vision becomes a bit more materialized? What value the current vision has for users?

Reply:
@reta , would it be fine to push this as a  module for 2.11 then ? Meanwhile, we are removing encryption-sdk from core.

Reply:
> @reta , would it be fine to push this as a module for 2.11 then ? Meanwhile, we are removing encryption-sdk from core : #9833

Thanks @gbbafna , we'll definitely get back to this once the 2.10 rush is over